### PR TITLE
Fix #10306 getServerHost

### LINKED ISSUE #10306: Jetty 12 generates wrong Host header
Jetty version: 12.0.0

Jetty environment: ee10 (embedded Jetty server configured to use HTTP/2)

Java version: openjdk version "20.0.1" 2023-04-18
OpenJDK Runtime Environment (Red_Hat-20.0.1.0.9-2.rolling.fc38) (build 20.0.1+9)
OpenJDK 64-Bit Server VM (Red_Hat-20.0.1.0.9-2.rolling.fc38) (build 20.0.1+9, mixed mode, sharing)

OS: Fedora 38, Ubuntu 23.04

Jetty returns the wrong host header when a request is forwarded from a different port. In my case the client sends a request to the default port 443 which is then forwarded to port 8443 which is the port Jetty listens on. The host header then incorrectly contains the port number 8443.

Server receives this (pardon the brackets):

request.getRequestURL().toString():

URL=[https://demo.jambo.software/demo.jambo.software/s;x=13y762q065p53vl8nbmb4uf720]

request.getHeader(...):

HEADER[sec-fetch-mode]=[cors]
HEADER[content-length]=[26]
HEADER[sec-fetch-site]=[same-origin]
HEADER[accept-language]=[en-US,en;q=0.9]
HEADER[cookie]=[13y762q065p53vl8nbmb4uf720=A25F3449-F86C-4545-805E-45F947465397]
HEADER[origin]=[https://demo.jambo.software/]
HEADER[Host]=[demo.jambo.software:8443]
HEADER[accept]=[*/*]
HEADER[sec-gpc]=[1]
HEADER[sec-ch-ua]=["Not/A)Brand";v="99", "Brave";v="115", "Chromium";v="115"]
HEADER[sec-ch-ua-mobile]=[?0]
HEADER[sec-ch-ua-platform]=["Linux"]
HEADER[content-type]=[application/vnd.piglet]
HEADER[accept-encoding]=[gzip, deflate, br]
HEADER[user-agent]=[Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36]
HEADER[sec-fetch-dest]=[empty]

While the netword tab of the browser gives this for the request:

Request URL:https://demo.jambo.software/demo.jambo.software/s;x=13y762q065p53vl8nbmb4uf720
Request Method:POST
Status Code:200
Remote Address:136.144.238.65:443
Referrer Policy:no-referrer

:authority:demo.jambo.software
:method:POST
:path:/demo.jambo.software/s;x=13y762q065p53vl8nbmb4uf720
:scheme:https
Accept:*/*
Accept-Encoding:gzip, deflate, br
Accept-Language:en-US,en;q=0.9
Content-Length:26
Content-Type:application/vnd.piglet
Cookie:13y762q065p53vl8nbmb4uf720=A25F3449-F86C-4545-805E-45F947465397
Origin:https://demo.jambo.software/
Sec-Ch-Ua:"Not/A)Brand";v="99", "Brave";v="115", "Chromium";v="115"
Sec-Ch-Ua-Mobile:?0
Sec-Ch-Ua-Platform:"Linux"
Sec-Fetch-Dest:empty
Sec-Fetch-Mode:cors
Sec-Fetch-Site:same-origin
Sec-Gpc:1
User-Agent:Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36

### PR DESCRIPTION
**Pull request by gregw**
Fix #10306 getServerHost

Use the default scheme before using the local port

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet @joakime The CI failure is a flake.  This is ready for review.

**Comment by gregw** [comment]
@sbordet nudge

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java [hunk 1]
```diff
@@ -77,10 +78,11 @@ static HostPort getServerAuthority(HttpConfiguration httpConfiguration, Connecti
...
 
-        SocketAddress local = connectionMetaData.getLocalSocketAddress();
-        if (local instanceof InetSocketAddress inet)
-            return new HostPort(inet.getHostString(), inet.getPort());
-
+        SocketAddress localSocketAddress = connectionMetaData.getLocalSocketAddress();
+        if (localSocketAddress instanceof InetSocketAddress inetSocketAddress)
+            return new HostPort(inetSocketAddress.getHostString(), inetSocketAddress.getPort());
+        else if (localSocketAddress != null)
+            return new HostPort(localSocketAddress.toString(), connectionMetaData.isSecure() ? 443 : 80);
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java]: Is this to support Unix-Domain?
If so, I am not sure we want to add an inet port to a non-inet host.

Also, method needs javadocs.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 2]
```diff
@@ -403,6 +403,13 @@ static int getRemotePort(Request request)
         return -1;
     }
 
+    /**
+     * Get the logical name for the server, which may differ from the actual server name
+     * @param request The request to get the server name of
+     * @return The logical name the request was sent to, which may be from the authority of the
+     *         request; the configured server authority; the actual network name of the server;
+     *         otherwise null
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: I'd prefer to keep `@return` short and simple, and add all the description and logic of what this method does to the main javadocs.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 3]
```diff
@@ -423,29 +430,38 @@ static String getServerName(Request request)
         return local == null ? null : local.toString();
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: Is it not this logic the same as `ConnectionMetaData getServerAuthority(HttpConfiguration httpConfiguration, ConnectionMetaData connectionMetaData)`?

I would call that, and then return `hostPort.getHost()`, so there is only 1 logic.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 4]
```diff
@@ -423,29 +430,38 @@ static String getServerName(Request request)
...
     }
 
+    /**
+     * Get the logical port a request was received on, which may differ from the actual network port.
+     * @param request The request to get the port of
+     * @return The port for the request, which may be from the authority of the request; the
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: Short `@return`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java (hunk 1 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java]: @sbordet so return null or a HostPort with no port?

Comment (sbordet) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java]: I would return null, the idea is that it is a last resort and most of the time there is an authority from somewhere else than the network.

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 5]
```diff
@@ -419,23 +418,18 @@ static String getServerName(Request request)
         if (uri.hasAuthority())
             return HostPort.normalizeHost(uri.getHost());
 
-        HostPort authority = request.getConnectionMetaData().getHttpConfiguration().getServerAuthority();
+        HostPort authority = request.getConnectionMetaData().getServerAuthority();
         if (authority != null)
             return HostPort.normalizeHost(authority.getHost());
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: No need to normalize again, as `HostPort` normalizes in the constructor.

**Review by sbordet — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 5 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java]: @sbordet Fixed, but will need a re-review to get green tick back

**Review by sbordet — APPROVED** [review]

**Pull request merged**